### ISSUE TITLE
Tonns of debug.log errors. What they means?

### ISSUE THREAD

**Issue by Krutoy242** [TheSilkMiner/Boson#2]
Just installed Boson to use math `log()` function. And it provided **hundreds** of `debug.log` errors.
I didn't change any configurations, all errors appeared with default configs.

Please, describe why there is so many errors, what they means for me and how to fix them if possible?

[debug.zip](https://github.com/TheSilkMiner/Boson/files/6158156/debug.zip)

There is list of all errors:

### 1. 48 Errors `Unable to transform class`
```java
[main/ERROR] [Fermion ASM/Transformer]: Unable to transform class 'net.thesilkminer.mc.fermion.test.asm.FermionTestSuite': no Transformers were accepted
```
What is transformers? What they transform?
I didnt carefully read whole documentation [there](https://docs.blamejared.com/1.12/en/Mods/Boson/Names/), maybe i missed what they do?

### 2. 11 Warnings `UNABLE TO CALL TRANSFORMER`
```java
[main/WARN] [Fermion ASM/Transformer]:         UNABLE TO CALL TRANSFORMER: It was disabled in the configuration file
```
What disabled? As i said before, i didnt disabled anything. Is it fine that default configuration erroring?

### 3. Warnings `Attempted to get`
```java
[Client thread/WARN] [Fermion/OtherClass Checks]: Attempted to get parameter of otherClass: expected '$ssuming $ is bills'; found ''
[Client thread/WARN] [Fermion/OtherClass Checks]: Attempted to get ID of otherClass: expected 3; found -1
[Client thread/WARN] [Fermion/OtherClass Checks]: Attempted to call 'print' of otherClass: expected 'Assuming a is bills>> 3 $ssuming $ is bills'; found ''
[Client thread/WARN] [Fermion/OtherClass Checks]: Attempted to call 'getId' of otherClass: expected 3; found -1
```
Whats going on there? What is `bills` and whats wrong with them?

### 4. Warnings `report the issue on our GitHub`
```java
[Client thread/WARN] [Fermion/OtherClass Checks]: If you've disabled the testing transformer, that's okay. If you haven't, report the issue on our GitHub!-1
```
Ok, reporting! 😄

### 5. Warnings `Attempting to register factory`
```java
[Client thread/WARN] [Boson/Recipes Processor]: Attempting to register factory 'smelting' with namespace 'minecraft' instead of the expected 'boson': this may cause errors later on!
```
Who registered? Where? I dont want any errors later!

### 6. 167 Warnings `This will NOT survive an upgrade`
```java
[Client thread/WARN] [Boson/Recipes Processor]: A mod has registered the factory 'avaritia:config' via code or assets. This will NOT survive an upgrade to 1.13! Start using data packs instead!
```
What upgrade? If this is about MC 1.13? Modpack planning to upgrade itself from 1.12 to 1.13? (is this even possible?).

### 7. 16 Errors `An error has occurred while attempting to process the file`
```java
[Client thread/ERROR] [Boson/Data-Pack Loader]: * An error has occurred while attempting to process the file 'spartanshields:shield_mod_bronze'.                                            *
```
This is not even file path (in Windows files cant have ":" in name). Who processing files and what purpose? Mod `Spartan Shields` works fine, why Boson shows errors then?

### 8. 111 WARNs `The mod '***' is defining some recipes in its 'assets/' directory`
```java
[Client thread/WARN] [Boson/Recipes in Assets]: The mod 'libvulpes' is defining some recipes in its 'assets/' directory: these will not survive a 1.13+ upgrade. Please consider moving them to 'data/' instead
```
Consider to moving assets? How? Im not mod developer, im just macking pack.

### 9. 2 ERRORs `Attempting to load a script while another one was loading!`
```java
[Client thread/ERROR] [Boson - CT Integration/Experimental Flags Handler]: * Attempting to load a script while another one was loading! Invariants have been broken! Attempting to correct: note that this may break further down the line
[Client thread/ERROR] [Boson - CT Integration/Experimental Flags Handler]: * Current Script: {[0:crafttweaker]: mods\RedstoneControl.zs}
[Client thread/ERROR] [Boson - CT Integration/Experimental Flags Handler]: * New file: mods\RFTools.zs
```
This files are fine! They have `#modloaded` define with mod names that was removed, so files should be just skipped. Why Boson dont like that?

### 10. 2 ERRORs `Unable to add '***' to tag`
```
[Client thread/ERROR] [Boson/Tag Loading Processor]: Unable to add 'minecraft:flowing_water' to tag 'minecraft:water', are you sure the entry exists? Error: Tags of type 'fluids' don't support null entries, but 'minecraft:flowing_water' was
[Client thread/ERROR] [Boson/Tag Loading Processor]: Unable to add 'minecraft:flowing_lava' to tag 'minecraft:lava', are you sure the entry exists? Error: Tags of type 'fluids' don't support null entries, but 'minecraft:flowing_lava' was

```
I didnt added anything in tags yet. Also, this is vanilla block. Is im in trouble so bad, even Water and Lava are wrong? 😅

**Comment by niklaswimmer** [comment]
Okay so let's go through this one by one

### 1. 48 Errors `Unable to transform class`
The reason you couldn't find something about transformers and what they do in the Docs is that they are actually not part of Boson and instead come with [Fermion](https://github.com/TheSilkMiner/Fermion/tree/1.12.2).
Regarding what they are and what they do: they transform java classes on byte code level, basically injecting new code at runtime into minecraft. 
The reason for those logs is that 1.12.2 loads coremods (a coremod consists out of several transformers) differently than 1.14.4 does, but because Fermion was initially developed for that newer version and the 1.12.2 version is a straight backport some things go out of order. Nevertheless this is not an issue and doesn't affect gameplay in any way. Unfortunately those error logs will exist and there is no way to turn them off/get rid of them.

### 2. 11 Warnings `UNABLE TO CALL TRANSFORMER`
This is just to inform that a transformer got disabled in the configuration file (in fact they are disabled by default). The reason we log disabled transformers is because in our opinion coremods should log about everything they do, so even that they do nothing. Another reason is so that mod developers can tell by the logs if someone disabled their transformers (which we think can reduce the time needed to find why something crashed).
So those can safely be ignored.

### 3. Warnings `Attempted to get`
So Fermion has some test transformers we run (during development) to ensure stuff works. Those test transformers are disabled in production (because they do nothing useful) but we forgot to remove our test logging. If Fermion ever gets an update for 1.12.2 we will likely remove those logs for production. For now you can either ignore it or remove the (error) logging by **enabling** following transformers in the config file located at `config/FermionEnv/fermion.asm.test/transformers.json`:
 - `test_runtime_field_at`
 - `test_runtime_method_at`

(When enabling the transformers the test passes and no error gets logged. Though we recommend just ignoring it)

### 4. Warnings `report the issue on our GitHub`
This belongs to Number 3 and basically says "if the test transformers are disabled ignore those logs". Well, as I just explained the test transformers are disabled.

### 5. Warnings `Attempting to register factory`
*Who registered?* - we (and by we I mean Boson) registered a recipe factory used to parse JSONs as smelting recipes and because smelting is a default vanilla thing we do it under minecraft's namespace instead of our own
*Where?* - Well, [here](https://github.com/TheSilkMiner/Boson/blob/1.12.2/src/main/resources/data/boson/recipes/_factories.json#L10) to be precisely
*Why?* - This allows mod and datapack makers to create smelting recipes through JSONs (since this is not possible by default)
*Errors?* - Nope, there won't be any

### 6. 167 Warnings `This will NOT survive an upgrade`
This log is to inform other modders that the way they do stuff will not work in newer versions of the game. Doing things the newer way reduces the amount of code needed to be rewritten to do a port.
In fact this log is useless for players (and modpack makers) which is why we are currently considering to make it optional and possible to disable. Please let us know if you would like to see an update with those logs removed/configurable. If wanted we could ship one in a couple of days.

### 7. 16 Errors `An error has occurred while attempting to process the file`
In 1.12.2 JSON recipes get put in the `asset/` folder. But because recipes are definitely not assets Boson allows recipes to be loaded from the `data/` folder (just like it's done in 1.13+). What happens here is that Spartan Shields has recipe files located in the `data/` folder which triggers Boson. Now the reason for the errors is probably because their recipes are formatted in 1.14 format (this could be the case if they have 1.14 stuff inside their 1.12 JAR - but we don't, and also can't, know that) but Boson tries to read JSON recipes in 1.14+ format (there was a format change between these versions) which fails. 
There is nothing we can do about this. Maybe it's better that the parsing fails because otherwise you would probably end up with unwanted recipes or completely different errors (for example when they try to register a recipe for an item which does not yet exist in their 1.12.2 version).

#### Update on that
After reading the full log message in your zip and examining Spartan Shields JAR File I can confirm that they have recipe files inside their `data/` folder which Boson tries to load. The reason for the error is that those recipes try to use a custom condition which is not specified inside the `data/` folder (so it's not the format as we first thought). 
Anyways, as already stated, there is nothing we can do about this.

`spartanshields:shield_mod_bronze` is the resource location of the file and translates to `data/spartanshields/recipes/shield_mod_bronze`

### 8. 111 WARNs `The mod '***' is defining some recipes in its 'assets/' directory`
This is the same as #6.

### 9. 2 ERRORs `Attempting to load a script while another one was loading!`
See [this issue on CraftTweakers GitHub](https://github.com/CraftTweaker/CraftTweaker/issues/984) for more information.

### 10. 2 ERRORs `Unable to add '***' to tag`
This in fact *is* an issue which we haven't noticed before, thanks for reporting. We will fix it in the next update.

**Comment by Krutoy242** [comment]
> if you would like to see an update with those logs removed/configurable

Yes, i would totally like.

Im not teacing you how to write mods, you probably already better programmers than me. But i want to share my oppinion about logging in Boson.

I dont have proper education, but [this article](https://sematext.com/blog/logging-levels/) tells, that
> ERROR – the log level that should be used when the application hits an issue *preventing functionalities from properly functioning*.

You say about `#1`
> doesn't affect gameplay in any way

This is wrong. If something doesn't affect anything it should come with `INFO` log level.

I believe that `ERROR` and `WARN` log levels exists to bring **attention** to problem. If not me nor you can do anything about `#7`, why log asking for attention?  
Also, please describe why Boson even parsing recipes on its own. I dont understand what functionality i have with this feature.

Error `#9` doesn't make sence.  
Its like developer added feature that doesn't work from the very beginning. Its wrong to notify players or modpack devs about this. Because this error would be like welcoming message for years (considering there is already year of issue). And welcoming messages never an Errors.

You said:
> coremods should log about everything they do, so even that they do nothing

I agree, but article said:
> WARN – the log level that indicates that something *unexpected* happened in the application, a problem, or a situation that might disturb one of the processes.

If functionality totally awared and by plan, why it is warning? It should be `INFO`.  
I feel like Boson is ☕ addict, and after 6th cup, even routine everyday stuff for him is WARNING 😄

About `#5`  
If Boson adding JSON recipe functionality, but it disabled, why it doing preparations? If nobody using this feature, and, if as you say
> since this is not possible by default

why classes are registered, and moves are done? I mean, isnt Minecraft would load faster if non-working code would not load at all?

About `#6`
> currently considering to make it optional and possible to disable

Please make it disabled by default. I wish **additional** functionality require **idditional** settings. The "informing modders" functionality is not even mentioned on Github page or CT docs "about" page.

----

In conclusion, its different how Boson working with logs from most of other mods.  
Usually, when i see `ERROR` or `WARN` in debug.log, this is reason to go into Google, issues trackers, ask in chats. I would like it Boson keep this attitude to errors.

**Comment by TheSilkMiner** [comment]
Let's tackle all of this in order.

### Log Levels
Log levels are always a bit open to interpretation, but what you said is also the standard I follow.

Now, "unable to transform class" is an actual `ERROR` since it prevents Fermion to actually fulfill its intended purpose, that is to transform classes.
Unfortunately, due to how the coremod loading process runs in 1.12.2, this is not fixable. Literally.
Forge says "provide an entry point that will load before your coremod loads", then fucking ignores it and loads the coremod. I am not able to work around that, so that's the current best solution.

As you're saying, if something doesn't affect gameplay it should come with `INFO` level. I beg to disagree as the following log messages shows (taken from your `debug.log`):

```
[19:14:04] [Client thread/ERROR] [tconstruct-modifier]: Cannot load modifier-model conarm:models/item/modifiers/antimagic_armor
java.io.FileNotFoundException: conarm:models/item/modifiers/antimagic_armor.json
[...]
[19:23:50] [Server thread/ERROR] [net.minecraft.advancements.AdvancementManager]: Parsing error loading built-in advancement minecraft:recipes/combat/diamond_chestplate
com.google.gson.JsonSyntaxException: Unknown recipe 'minecraft:diamond_chestplate'
```

These are messages that signal errors, but at the same time, the game carries on just fine without those anyway.

`WARN`s on the other hand indicate that something unexpected happen, but stuff may continue just fine. This is the case for most warnings in Boson.
They are warnings for a reason: they indicate something is weird, but that something may very well be intended. As an example, let's consider this from your logs:

```
[19:16:24] [Client thread/WARN] [FML]: Potentially Dangerous alternative prefix `minecraft` for name `pampresser`, expected `harvestcraft`. This could be a intended override, but in most cases indicates a broken mod.
```

Nevertheless, I agree that log spam in large modpacks may not be desirable, so the config option has been added: `common/recipes/suppress_recipe_warnings`. You can set that to off to suppress recipe-loading related ones.

### Datapack Loading
Datapack loading is one of the main features of Boson, if not the main one I'd say. Tags, recipes, loaders... all of those are based around the datapack loading mechanism that Boson has.
And datapacks have been backported down from newer Minecraft versions (though I admit they're less powerful than I'd like simply because mods don't really work with those, I can't really blame them though).
This is why Boson loads files in `data` directories or adds a JSON factory to register smelting recipes, or blasting ones, or campfire ones. Only some of those are supported (no campfires exist in 1.12.2, for example).
Nevertheless, Boson makes an attempt in being as compatible as possible with datapacks, which means setting up and loading the needed files. This is also why Tags are a thing, BTW: to allow more powerful and data-driven ore dictionary.

### Error 1
That's the result of using old stuff, and it's not really fixable. You'd be seeing errors in one place or the other.

### Error 2
I am going to publish a test-less version of Fermion, which will stop tests from loading in production. They should have never been there, and were left as a debugging tool. I guess that people that still want them can simply downgrade.

### Error 3
See Error 2

### Error 4
See Error 2

### Error 5
Non-working code? That code works just fine. It simply enables another way of registering smelting recipes instead of using raw code. And this ties into datapack loading that has been mentioned before.
It's a `WARN`, meaning that something might be wrong. It's not, it can be ignored, but the datapack has no way of telling so it notifies you, the user.

### Error 6
The config option has been added.

### Error 7
That's just a datapack loading and a mod shipping one. Such datapack is not working so it was probably not intended to be there. But it is, so...
A bit like if you were to ship a script that caused an error because of a missing semicolon. CraftTweaker is going to complain, even if you didn't mean to leave the script in the pack.

### Error 8
See Error 6

### Error 9
That error absolutely makes sense. CraftTweaker is telling me that a script has started loading, but it never finished. Like, if you were to start reading a book but never stop.
Obviously this is a problem, especially if Boson's CraftTweaker integration relies on knowing when files load and unload to set up some additional settings (experimental flags, as an example).
The error is Boson scrambling and attempting to try and guess what is going on. The alternative would be to simply shut down all file-based CT integration, and nobody wants that.
And since Boson is basically the only mod I know of, if not the only one in existence, that relies on this events, it kinda makes sense that nobody else reported the issue.

### Error 10
Fixed, thanks for the report. I don't know why, I literally never noticed.

**Comment by Krutoy242** [comment]
Ok, seems like you answered on all my complains.
This issue is fulfilled i guess.

P.S. I would like to read documentation as for modpack maker (not for mod dev). I feel like Boson has a lot of features, that do something useful, but for me they only erroring. I want to use them! Datapacks for example - have no clue what is that.

**Comment by TheSilkMiner** [comment]
Fixed versions will be on CurseForge, currently waiting on approval. As for data packs, refer to [the official wiki](https://minecraft.fandom.com/wiki/Data_Pack). Boson had along its goals the backporting of some those features, so not **everything** will be available, but some stuff is.
For more information, feel free to contact me over at the MPD or BlameJared guilds on Discord, since GitHub Issues isn't the best place to describe usage features 😜.
